### DISCUSSION TITLE
How to call existing C# function (from an existing `.cs` file) from Python?

### DISCUSSION THREAD

**Discussion by vadimkantorov** [pythonnet/pythonnet#2195] [Q&A] [answered]
At https://pythonnet.github.io/pythonnet/python.html there is no example of calling a user C# function from a C# file.

Must we first manually compile the C# code to an assembly and then load it with `import clr; clr.AddReference('my_function_in_assembly.dll')`? Or can PythonNet itself call the Roslyn compiler at runtime? If so, it would be great to have a complete example in docs.

Being able to call a C# function from Python is an important scenario for a soft migration from C# code to Python code.

**Comment by filmor** [comment]
This is not directly supported, you'll have to call the respective functions yourself (maybe try.dot.net is a good start, but I can't find the part where they are doing the compilation right now). I also don't really understand how this would simplify a "soft migration", as this is not how you'd usually run C# code...

**Reply by vadimkantorov** [reply]
I've figured some things out :)

I created a file `foo.cs`
```csharp
public class Foo
{
    public string Hi(string x)
    {
        return "Hi " + x;
    }

    public static string Hello(string x)
    {
        return "Hello " + x;
    }
}
```

Installed .NET Core 7.0 and compiled `foo.cs` into `foo.dll` as follows:
```shell
#! /bin/bash

# https://learn.microsoft.com/en-us/dotnet/core/install/linux-scripted-manual#scripted-install
# https://stackoverflow.com/a/65283791/445810

#wget https://dot.net/v1/dotnet-install.sh -O dotnet-install.sh
#chmod +x ./dotnet-install.sh
#./dotnet-install.sh --channel 7.0

DOTNET_HOME=$HOME/.dotnet
DOTNETSDKVER=$("$DOTNET_HOME/dotnet" --version)
DOTNETFWKVER=$("$DOTNET_HOME/dotnet" --list-runtimes | grep Microsoft.NETCore.App | tail -n 1 | tail -n 1 | cut -d' ' -f2)
DOTNETLIBDIR="$DOTNET_HOME/shared/Microsoft.NETCore.App/$DOTNETFWKVER"
"$DOTNET_HOME/dotnet"  "$DOTNET_HOME/sdk/$DOTNETSDKVER/Roslyn/bincore/csc.dll" $(find "$DOTNETLIBDIR" -name "*.dll" | xargs printf '-r:"%s" ')  -target:library -out:foo.dll foo.cs
```

And managed to compile it into a `foo.dll`

Then the following code executes successfully:
```python
import os

os.environ['DOTNET_ROOT'] = os.path.expanduser('~/.dotnet')

import clr_loader
import pythonnet

pythonnet.set_runtime(clr_loader.get_coreclr())

import clr
clr.AddReference(os.path.abspath('foo.dll'))
print(clr.Foo().Hi('world'))
print(clr.Foo.Hello('you'))
``` 
runs successfully!

although `from foo import Foo`. Is it expected? This is a bit more explicit and seems a bit better.

Feel free to add this example to the docs :)

**Reply by vadimkantorov** [reply, marked as answer]
Here is my complete example gist: https://gist.github.com/vadimkantorov/c9ab97aff995ea01ea7f6248ddc733bc

**Reply by vadimkantorov** [reply]
I was basically not sure if modern .NET Core included some sort of interpreter (CSharpScript-like) which can consume C# code directly without first manually compiling it into a DLL on disk and if Python.NET included interop with this interpreter Roslyn tech

**Reply by vadimkantorov** [reply]
Although, maybe another good example would be an example of calling the Roslyn compiler `csc.dll` from using `pythonnet` directly. Without having to do it from the shell script.
